### PR TITLE
k-step arnoldi factorization

### PR DESCRIPTION
**Pull request by mganahl**

### CODE REVIEW

File: tensornetwork/backends/jax/jitted_functions.py [hunk 1]
```diff
@@ -149,3 +150,146 @@ def body_vector(i, vals):
...
+        return jax.lax.cond(val < thresh, False, lambda x: x, True, lambda x: x)
+
+      return jax.lax.cond(iteration < maxiter, [norm, threshold], check_thresh,
+                          False, lambda x: x)
+
+    norms_dtype = np.real(v0.dtype).dtype
+    kvfinal, Hfinal, _, _, norm, _, it, _ = jax.lax.while_loop(
+        cond_fun, body, [
+            krylov_vectors, H, matvec, v,
+            norms_dtype.type(100000000000.0), eps, start, num_krylov_vecs
```
Comment (Thenerdstation) [inline tensornetwork/backends/jax/jitted_functions.py:291]: any reason for the magic bajillion here?

Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py:291]: just any large positive number should dp

Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py:291]: its the initial  value of a convergence flag

Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py]: I added better formatting

File: tensornetwork/backends/jax/jitted_functions.py [hunk 2]
```diff
@@ -59,9 +59,8 @@ def jax_lanczos(matvec, arguments, init, ncv, neig, landelta, reortho):
     def body_modified_gram_schmidt(i, vals):
       vector, krylov_vectors = vals
       v = krylov_vectors[i, :]
-      vector -= jax.numpy.dot(jax.numpy.conj(v),
-                              jax.numpy.ravel(vector)) * jax.numpy.reshape(
-                                  v, vector.shape)
+      vector -= jax.numpy.vdot(v, jax.numpy.ravel(vector)) * jax.numpy.reshape(
+          v, vector.shape)
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions.py:63]: Weren't we using jnp? `ravel` is also redundant inside `vdot`. What about

`vector -= jnp.vdot(v, vector).reshape(vector.shape)`

File: tensornetwork/backends/jax/jitted_functions.py [hunk 3]
```diff
@@ -149,3 +150,151 @@ def body_vector(i, vals):
...
+    m = min(`it`,`num_krylov_vecs`). The factorization will 
+    do at most `num_krylov_vecs` steps. The returned arrays 
+    `kv` and `H` will satisfy the Arnoldi recurrence relation
+    ```
+    matrix @ Vm - Vm @ Hm - fm * em = 0
+    ```
+    with `matrix` the matrix representation of `matvec` and 
+    `Vm =  jax.numpy.transpose(kv[:it, :])`, 
+    `Hm = H[:it, :it]`, `fm = np.expand_dims(kv[it, :] * H[it, it - 1]`,1) 
+    and `em` a kartesian basis vector of shape `(1, kv.shape[1])`
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions.py:239]: typo

File: tensornetwork/backends/jax/jitted_functions.py [hunk 4]
```diff
@@ -149,3 +150,151 @@ def body_vector(i, vals):
...
+        `krylov_vectors`: jax.array of collected krylov vectors
+        `n`: integer denoting the column-position of the overlap 
+          <`krylov_vector`|`vector`> within `H`.
+    Returns:
+      updated vals.
+             
+    """
+    vector, krylov_vectors, n, H = vals
+    v = krylov_vectors[j, :]
+    h = jax.numpy.vdot(v, jax.numpy.ravel(vector))
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions.py:220]: Redundant `ravel`

File: tensornetwork/backends/jax/jitted_functions.py [hunk 5]
```diff
@@ -149,3 +150,151 @@ def body_vector(i, vals):
...
+    H = jax.ops.index_update(H, jax.ops.index[j, n], h)
+    vector = vector - h * jax.numpy.reshape(v, vector.shape)
+    return [vector, krylov_vectors, n, H]
+
+  @partial(jax.jit, static_argnums=(5, 6, 7))
+  def _arnoldi_fact(matvec, args, v0, krylov_vectors, H, start, num_krylov_vecs,
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions.py:226]: We should probably settle on what we are going to call the number of Krylov vectors in these various algorithms; in Lanczsos it looks like you call the same thing `ncv`. I personally like either that or `n_kry`; `num_krylov_vecs` feels really long to me but it's a matter of taste.

File: tensornetwork/backends/jax/jitted_functions.py [hunk 6]
```diff
@@ -149,3 +150,151 @@ def body_vector(i, vals):
...
+                                            jax.ops.index[i + 1, :],
+                                            jax.numpy.ravel(Av))
+      return [krylov_vectors, H, matvec, Av, norm, threshold, i + 1, maxiter]
+
+    def cond_fun(vals):
+      _, _, _, _, norm, threshold, iteration, maxiter = vals
+
+      def check_thresh(check_vals):
+        val, thresh = check_vals
+        return jax.lax.cond(val < thresh, False, lambda x: x, True, lambda x: x)
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions.py:291]: These might deserve a block comment spelling out what they do? Not critical though.

File: tensornetwork/backends/jax/jitted_functions_test.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+from tensornetwork.backends.jax.jitted_functions import _generate_arnoldi_factorization
+jax.config.update('jax_enable_x64', True)
+
+
+def test_arnoldi_factorization():
+  np.random.seed(10)
+  D = 20
+  mat = np.random.rand(D, D)
+  x = np.random.rand(D)
+  dtype = np.float64
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions_test.py:14]: Is it sufficient to just test the one dtype?

File: tensornetwork/backends/jax/jitted_functions.py (hunk 2 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py:63]: jnp is only used within the JaxBackend

File: tensornetwork/backends/jax/jitted_functions.py (hunk 5 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py:226]: You're right, we should be consistent here. 

I think long argument names are generally the preferable choice, in particular in backend modules that don't interface with the users.

Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py:226]: or let me rephrase: expressive argument names are preferable

File: tensornetwork/backends/jax/jitted_functions_test.py (hunk 7 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions_test.py:14]: no, good catch!

File: tensornetwork/backends/jax/jitted_functions.py (hunk 3 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py:239]: you mean kartesian?

File: tensornetwork/backends/jax/jitted_functions.py (hunk 2 above)
Comment (alewis) [reply tensornetwork/backends/jax/jitted_functions.py]: Fair enough. `vdot` still flattens by definition, though.

File: tensornetwork/backends/jax/jitted_functions.py (hunk 3 above)
Comment (alewis) [reply tensornetwork/backends/jax/jitted_functions.py]: That is what I was referring to, but if you were writing that on purpose for some reason go for it

File: tensornetwork/backends/jax/jitted_functions.py (hunk 5 above)
Comment (alewis) [reply tensornetwork/backends/jax/jitted_functions.py:226]: I suppose `num_krylov_vecs` is fine, then

File: tensornetwork/backends/jax/jitted_functions.py (hunk 2 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py]: removed `ravel`

File: tensornetwork/backends/jax/jitted_functions.py (hunk 3 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py]: fixedx

File: tensornetwork/backends/jax/jitted_functions.py (hunk 5 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py:226]: i usually use `ncv` for user-facing functions to keep it consistent with scipy. Is taht OK?

**Comment by alewis** [comment]
Yeah, that makes sense to me. The difficulty is that SciPy itself is not
consistent: this parameter gets called ncv in eigh, for example, and
"restart" (which I find very confusing) in gmres, for example. When I went
through this thought process writing vumps, I eventually called it n_krylov
throughout on the grounds that consistency wasn't possible and neither of
the original names are very expressive. ncv consistently is ok too with me

**Comment by Thenerdstation** [comment]
Yeah, I vote `ncv` to be consistent with scipy.

**Comment by mganahl** [comment]
I left it `num_krylov_vecs` for now because I'll have to change this for all backends and in all tests. This is better to submit in a separate PR

File: tensornetwork/backends/jax/jitted_functions.py [hunk 8]
```diff
@@ -115,7 +111,9 @@ def check_thresh(check_vals):
     diag_elems = jax.numpy.zeros(ncv, dtype=init.dtype)
 
     norms = jax.ops.index_update(norms, jax.ops.index[0], 1.0)
-    norms_dtype = np.real(init.dtype).dtype
+
+    norms_dtype = jax.numpy.real(jax.numpy.empty((0, 0),
+                                                 dtype=init.dtype)).dtype
```
Comment (Thenerdstation) [inline tensornetwork/backends/jax/jitted_functions.py:116]: revert this.

File: tensornetwork/backends/jax/jitted_functions.py [hunk 9]
```diff
@@ -149,3 +147,163 @@ def body_vector(i, vals):
...
+  with `matrix` the matrix representation of `matvec` and 
+  `Vm =  jax.numpy.transpose(kv[:it, :])`, 
+  `Hm = H[:it, :it]`, `fm = np.expand_dims(kv[it, :] * H[it, it - 1]`,1) 
+  and `em` a kartesian basis vector of shape `(1, kv.shape[1])` 
+  with `em[0, -1] == 1` and 0 elsewhere.
+
+  Note that the caller is responsible for dtype consistency between 
+  the inputs, i.e. dtypes between all input arrays have to match.
+
+  The arguments to `_arnoldi_fact` are:
```
Comment (Thenerdstation) [inline tensornetwork/backends/jax/jitted_functions.py:175]: Remove this line.

File: tensornetwork/backends/jax/jitted_functions.py (hunk 8 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py:116]: the old code broke on my system

Comment (Thenerdstation) [reply tensornetwork/backends/jax/jitted_functions.py:116]: Why? What was the error?

Doing it this way is very hacky. There must be a better way.

Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py:116]: it works now so I reverted it.
